File: problems/238/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        l = len(nums)
        ans = [0] * l
        ans[0] = 1
        
        for i in range(1, l):
            ans[i] = ans[i - 1] * nums[i - 1]
        
        r = 1
        
        for i in reversed(range(l)):
            ans[i] = ans[i] * r
            r *= nums[i]
        
        return ans

File: problems/238/solution.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = size(nums);
        vector<int> ans(n, 1);
        
        for (int prev = 1, i = 0; i < n; i++) {
            ans[i] = prev;
            prev *= nums[i];
        }
        
        for (int prev = 1, i = n - 1; i >= 0; i--) {
            ans[i] *= prev;
            prev *= nums[i];
        }
        
        return ans;
    }
};

/*

1  2  3  4
1  1  2  6
24 12 4  1
24 12 8  6

*/